src/authn/session.js

```diff
@@ -0,0 +1,84 @@
+const {keyringFromParamString, encrypt, decrypt, randomKey} = require("../cryptography");
+
+
+const PRODUCTION = process.env.NODE_ENV === "production";
+
+const SESSION_ENCRYPTION_KEYS = process.env.SESSION_ENCRYPTION_KEYS;
+
+if (PRODUCTION && !SESSION_ENCRYPTION_KEYS) {
+  throw new Error("SESSION_ENCRYPTION_KEYS required in production mode");
+}
+
+const KEYRING = keyringFromParamString(SESSION_ENCRYPTION_KEYS || `RANDOM=${randomKey()}`);
+
+
+/**
+ * Decrypts and returns the encrypted tokens stored in the given session
+ * object.
+ *
+ * If there are no tokens in the session, then returns null.
+ *
+ * @param {Object} session - typically req.session
+ * @returns {{idToken, accessToken, refreshTokens}|null}
+ */
+async function getTokens(session) {
+  if (session?.encryptedTokens) {
+    const context = {sid: session.id};
+    const decryptedTokens = JSON.parse(await decrypt(KEYRING, session.encryptedTokens, context));
```

Comment:
Should we catch the assertion error for mismatched encryption context here and just return null if the decryption fails?

Reply:
For similar reasons as above to error rather than handle and move on, I'd rather throw a big loud error if the context doesn't match, as this means either a bug (most likely) or, worse, someone trying to reuse encrypted material in the wrong context, which would mean much larger problems.

Reply:
Ah, sounds like we need a better logging/monitoring system for server issues (I know you've mentioned this a lot...)
I just didn't think we'd want one bad actor to take down the entire server.

Reply:
Oh. It won't take down the server, just tank the request with a 500 error. (But yes, I'd love [better logging](https://github.com/nextstrain/nextstrain.org/issues/511).)